src/rdiff_backup/rpath.py

```diff
@@ -997,8 +999,26 @@ def get_temp_rpath(self, sibling=False):
             if self.__class__._temp_file_index > 100000000:
                 log.Log("Warning: Resetting tempfile index", 2)
                 self.__class__._temp_file_index = 0
-            tf = self.append('rdiff-backup.tmp.{index:d}'.format(
-                             index=self.__class__._temp_file_index))
+
+            # When the file system hosting the rdiff-backup-data directory
+            # is full and when the --tempdir flag is defined, attempt to save
+            # temporary files on a different path / file system.
+            if tempfile.tempdir and (shutil.disk_usage(self.path).free == 0):
+                # If tempfile.tempdir is manually passed in via the --tempdir
+                # cli flag, it defaults being a bytes string, as such we need to
+                # convert the target path to a string first
+                if type(tempfile.tempdir) == bytes:
```

Comment:
Sorry, if I wasn't clear but you don't need to check for bytes vs. str, `os.fsdecode` does it for you, so you can replace the four lines just with one `tempdir = os.fsdecode(tempfile.tempdir)` and it will work even if `tempfile.tempdir` is already a string. See https://docs.python.org/3/library/os.html#os.fsdecode

Reply:
For a moment I missed the fact os.fsdecode() would cope just fine with both a byte string and a string type, defaulting to return a string, change made, thanks!